internal/app/server/open_saves.go

```diff
@@ -621,6 +621,44 @@ func (s *openSavesServer) CommitChunkedUpload(ctx context.Context, req *pb.Commi
 	return blob.ToProto(), nil
 }
 
+func (s *openSavesServer) CommitChunkedUploadWithUpdateRecord(ctx context.Context, req *pb.CommitChunkedUploadWithUpdateRecordRequest) (*pb.BlobMetadata, error) {
+	storeKey := req.GetUpdateRecord().GetStoreKey()
+	pbRecord := req.GetUpdateRecord().GetRecord()
+	blobKey, err := uuid.Parse(req.GetSessionId())
+	if err != nil {
+		log.Errorf("SessionId is not a valid UUID string: %v", err)
+		return nil, status.Errorf(codes.InvalidArgument, "SessionId is not a valid UUID string: %v", err)
+	}
+	blob, err := s.metaDB.GetBlobRef(ctx, blobKey)
+	if err != nil {
+		log.Errorf("Cannot retrieve chunked blob metadata for session (%v): %v", blobKey, err)
+		return nil, err
+	}
+	updateTo, err := record.FromProto(storeKey, pbRecord)
+	if err != nil {
+		log.Errorf("Invalid proto for store (%s), record (%s): %v", storeKey, pbRecord.GetKey(), err)
+		return nil, status.Errorf(codes.InvalidArgument, "Invalid record proto: %v", err)
+	}
+	record, _, err := s.metaDB.PromoteBlobRefWithRecordUpdater(ctx, blob, updateTo, func(r *record.Record) (*record.Record, error) {
+		r.BlobSize = blob.Size
+		r.ExternalBlob = blob.Key
+		r.Chunked = blob.Chunked
+		r.Timestamps.Update()
+		r.OwnerID = updateTo.OwnerID
+		r.Properties = updateTo.Properties
+		r.Tags = updateTo.Tags
+		r.OpaqueString = updateTo.OpaqueString
+		return r, nil
+	})
+	if err != nil {
+		log.Errorf("PromoteBlobRefWithRecordUpdater failed for object %v: %v", blob.ObjectPath(), err)
+		// Do not delete the blob object here. Leave it to the garbage collector.
+		return nil, err
+	}
+	s.cacheRecord(ctx, record, req.GetHint())
```

Comment:
Here we use the hint from the request, but there is also a hint in the UpdateRecordRequest. See comments above about solving this ambiguity.